File: daa3.java
import java.util.*;

class Item{
    double val, weight;

    Item(double val, double weight){
        this.val = val;
        this.weight = weight;
    }
}

public class daa3 {
    
    static Comparator <Item> cmp = (a,b)->{
        double r1 = a.val/a.weight;
        double r2 = b.val/b.weight;
        return Double.compare(r2,r1); // r2 before r1 for descending
    };

    static double fracKnap(int n, int W, List<Item> items){
        
        Collections.sort(items,cmp);
        double totalValue = 0;

        for(int i=0; i<n; i++){
            if (items.get(i).weight < W){
                totalValue += items.get(i).val;
                W -= items.get(i).weight;
            }
            else{
                totalValue += ( W * (items.get(i).val /items.get(i).weight) );
                break; 
            }
        }

        return totalValue;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        System.out.println("enter total no. of items: ");
        int n = sc.nextInt();
        
        List <Item> items = new ArrayList<>();
        System.out.println("enter value and weight of each item: ");

        for(int i=0; i<n; i++){
            double val = sc.nextDouble();
            double weight = sc.nextDouble();
            items.add(new Item(val, weight));
        }

        System.out.println("enter capacity of knapsack: ");
        int W = sc.nextInt();

        double maxValue = fracKnap(n, W, items);
        System.out.println("max value possible in knapsack: " + maxValue);

        sc.close();
    }
}

File: daa3.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;
float fractionalKnapsack(vector<vector<float>>&items , float capacity){
    int n = items.size();
    for(int i=0; i<n;i++){
        items[i][2]= items[i][0]/items[i][1];
    }

    sort(items.begin(),items.end(),[](vector<float>&a,vector<float>&b){
              return a[2]>b[2];
    });
    float totalProfit = 0.0;
    for(int i=0;i<n;i++){
        if(capacity>=items[i][1]){
            totalProfit += items[i][0];
            capacity -= items[i][1];
        }
        else{
            totalProfit += items[i][2] * capacity;
            break;
        }
    }
    return totalProfit;
}
int main(){
    int n;
    cout<<"Enter the number of Objects : ";
    cin>>n;

    float capacity;
    cout<<"Enter the knapsack capacity : ";
    cin>>capacity;

    vector<vector<float>>items(n,vector<float>(3));
    
    cout << "Enter profit and weight of each item:\n";
    for(int i=0;i<n;i++){
        cin>>items[i][0]>>items[i][1];
    }

    double maxProfit = fractionalKnapsack(items,capacity);
    cout<<"MaxProfit : "<<maxProfit;
    return 0;
}

// output - 
// Enter the number of Objects : 7
// Enter the knapsack capacity : 15
// Enter profit and weight of each item:
// 12 2
// 17 5
// 9 7
// 7 3
// 5 1
// 21 4
// 8 1
// MaxProfit : 67.6667